Implement queue-ordered memory buffers

fwyzard opened the pull request:
Move the implementation of the allocation and deallocation strategy from the `BufCpuImpl` and `BufUniformCudaHipRt` classes to the `AllocBuf<DevCpu>` and `AllocBuf<DevUniformCudaHipRt>` factory functors. This allows the implementation of different strategies in the `allocBuf` or other functions, without affecting the buffer classes.

Add the `allocAsyncBuf` function to allocate buffers using non-blocking, queue-ordered memory allocations and deallocations.

These functions take as argument a queue instead of a device; the memory associated with the buffer is guaranteed to be "alive" according to the lifetime of the buffer as seen by the queue: from the moment the queue reaches the buffer's constructor to the moment it reaches the destructor.

So far only the `DevCpu` and `DevCudaRt` devices are supported.
Attempting to use this with HIP raises an error, since HIP does not support the underlying stream-ordered memory allocation functions.

The `DevCpu` implementation
  - allocates memory immediately;
  - frees memory when the queue associated to the buffer reaches the point where the destructor of the buffer is called.

The `DevCudaRt` implementation
  - allocates memory with `cudaMallocAsync`, so only 1D buffers are supported;
  - frees memory with `cudaFreeAsync`.

fwyzard committed:
Implement queue-ordered memory buffers

Add the allocAsyncBuf function to allocate buffers using non-blocking,
queue-ordered memory allocations and deallocations.

These functions take as argument a queue instead of a device; the memory
associated with the buffer is guaranteed to be "alive" according to the
lifetime of the buffer as seen by the queue: from the moment the queue
reaches the buffer's constructor to the moment it reaches the destructor.

So far only the DevCpu and DevCudaRt devices are supported.
Attempting to use this with HIP raises an error, since it does not
support the underlying stream-ordered memory allocation functions.

The DevCpu implementation
  - allocates memory immediately;
  - frees memory when the queue associated to the buffer reaches the point
    where the destructor of the buffer is called.

The DevCudaRt implementation
  - requires CUDA 11.2 or later;
  - allocates memory with cudaMallocAsync, so only 1D buffers are supported;
  - frees memory with cudaFreeAsync.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard committed:
Move the allocation logic outside of the BufCpuImpl class

Move the allocation and deallocation from the BufCpuImpl class to the
AllocBuf<DevCpu> factory functor.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard committed:
Move the allocation logic outside of the BufUniformCudaHipRt class

Move the allocation and deallocation from the BufUniformCudaHipRt class
to the AllocBuf<DevUniformCudaHipRt> factory functor.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
This is a reimplementation of #1480 following the discussion in https://github.com/alpaka-group/alpaka/pull/1480#issuecomment-977975403 and later.

fwyzard commented on include/alpaka/mem/buf/Traits.hpp:
```diff
@@ -69,6 +77,25 @@ namespace alpaka
         return traits::BufAlloc<TElem, Dim<TExtent>, TIdx, TDev>::allocBuf(dev, extent);
     }
 
+    //! The stream-ordered memory buffer type trait alias template to remove the ::type.
+    template<typename TDev, typename TElem, typename TDim, typename TIdx>
+    using AsyncBuf = typename traits::AsyncBufType<alpaka::Dev<TDev>, TElem, TDim, TIdx>::type;
```
I'm not convinced that this is needed with this approach, since the `AsyncBufType` should always be the same as the `BufType` when it is defined.
On the other hand, since currently this is only defined for the CPU and CUDA devices, it could be used to check if the async buffer functionality is actually implemented for a device ?

bernhardmgruber replied:
So far it seems we don't need it. I would remove it.

fwyzard replied:
OK.

fwyzard replied:
Maybe add a default implementation for `AsyncBufAlloc` that fails with a `static_assert` and informs the user that the given buffer does not support stream-ordered memory allocation ?

fwyzard replied:
Or better to leave it undefined, so it can be checked with SFINAE ?

bernhardmgruber replied:
Wouldn't that eliminate the purpose of alpaka, to give you a best effort implementation of a particular behavior across all accelerators? Your code would no longer be portable. And we can workaround the limitations on all backends by emplacing a host task into the queue, that frees the buffer.

In case you could query though, whether the underlying backend supports async alloc/free, what would you do with the information?

fwyzard replied:
> And we can workaround the limitations on all backends by emplacing a host task into the queue, that frees the buffer.

I agree with you.
Sorry, I didn't mean that one should _not_ implement the stream-ordered behaviour all accelerator types.
But this PR currently does not, and I thought it would be useful so make it explicit somehow.

I'm happy to remove the `AsyncBuf` trait and _not_ add any checks, with the goal to eventually support stream-ordered allocations on every accelerator.

> In case you could query though, whether the underlying backend supports async alloc/free, what would you do with the information?

I guess the user code could fall back to some private implementation, or adapt the algorithms to use the standard allocations with some extra synchronisations.

bernhardmgruber replied:
IIUC only the HIP, OpenAcc and OpenMP5 backends are missing? Would it be too hard to provide a quick implementation for them? With a callback, similar to the CPU implementation? If it is, I could open an issue and implement it later.

bernhardmgruber replied:
> I guess the user code could fall back to some private implementation, or adapt the algorithms to use the standard allocations with some extra synchronisations.

Hmm. I asked the question because I wondered if the user would then resort to anyway implement the async alloc/free themselves by placing a callback into the queue. Phrased differently: can the user actually implement something better than our fallback implementation of placing the callback. Maybe we should provide a query than. But we can also wait until someone actually needs this feature.

fwyzard replied:
I don't have experience with OpenACC and OpenMP5, so I'm not sure about those :-(

For HIP I don't think there is a CPU thread managed by Alpaka associated with the Queue, so we cannot use the same approach used for the PU buffer.
And, assuming it has the same limitations as CUDA that a callback cannot call a CUDA/HIP function, a callback cannot be used either.
The other implementation I can think of is a full-fledged caching memory allocator; that's my long term goal anyway, but not something I can implement quickly. Also, so far it's not clear if it would be desirable to have it inside Alpaka or built on top or on the side.

bernhardmgruber replied:
> I don't have experience with OpenACC and OpenMP5, so I'm not sure about those :-(

Let's open an issue then when this PR is merged. @jkelling is experienced with those backends, so he might have a look later.

> And, assuming it has the same limitations as CUDA that a callback cannot call a CUDA/HIP function, a callback cannot be used either.

I think we have a workaround for that by keeping a background thread around. But I guess we can also implemented that later.

fwyzard commented on include/alpaka/mem/buf/BufUniformCudaHipRt.hpp:
```diff
@@ -89,21 +85,8 @@ namespace alpaka
-        }
-
     public:
-        DevUniformCudaHipRt m_dev; // NOTE: The device has to be destructed after the memory pointer because it is
-                                   // required for destruction.
```
The comment is no longer relevant, because the deleter captures the device by value.

fwyzard commented on include/alpaka/mem/buf/BufCpu.hpp:
```diff
@@ -81,24 +79,12 @@ namespace alpaka
             DevCpu const m_dev;
             Vec<TDim, TIdx> const m_extentElements;
-            TElem* const m_pMem;
+            std::shared_ptr<TElem> const m_spMem;
```
This approach does add a small overhead, replacing a `TElem*` with a `std::shared_ptr<TElem>`.
This could be avoided by storing the deleter directly in an `std::function` instead of using the `shared_ptr` for that.

An other possibility would be to follow the same approach used by `BufUniformCudaHipRt`:
  - move the metadata (device, extent, pitch) from `BufCpuImpl` directly into `BufCpu`
  - replace the `shared_ptr<BufCpuImpl>` with a `shared_ptr<TElem>`, and set the destructor there

However, this would trade the one less indirection with more expensive copies for the `BufCpu` objects, since the metadata would have to be copied as well.

fwyzard replied:
Storing the deleter explicitly proved to be simple enough, and should add a smaller overhead than using an extra `shared_ptr`.

fwyzard commented on include/alpaka/mem/buf/BufCpu.hpp:
```diff
@@ -82,23 +82,14 @@ namespace alpaka
             DevCpu const m_dev;
             Vec<TDim, TIdx> const m_extentElements;
             TElem* const m_pMem;
+            std::function<void(TElem*)> m_fDeleter;
```
is `m_deleter` a better choice of name ?

bernhardmgruber replied:
Yes please. The `f` reminds me too much of Hungarian notation.

fwyzard commented:
I don't understand this error [from the CI](https://github.com/fwyzard/alpaka/runs/4321763635?check_suite_focus=true#step:5:5135):
```
 /home/runner/work/alpaka/alpaka/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp:396:23: error: no member named 'DependentFalseType' in namespace 'alpaka::meta'
                meta::DependentFalseType<TElem>::value,
                ~~~~~~^
/home/runner/work/alpaka/alpaka/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp:396:42: error: 'TElem' does not refer to a value
                meta::DependentFalseType<TElem>::value,
                                         ^
/home/runner/work/alpaka/alpaka/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp:392:27: note: declared here
        template<typename TElem, typename TDim, typename TIdx>
                          ^
/home/runner/work/alpaka/alpaka/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp:396:50: error: no member named 'value' in the global namespace
                meta::DependentFalseType<TElem>::value,
                                               ~~^
3 errors generated when compiling for sm_30.
```

bernhardmgruber commented:
Are you missing an `#include <alpaka/meta/DependentFalseType.hpp>` in `BufUniformCudaHipRt.hpp`?

fwyzard commented:
Ah, thanks, good catch !

bernhardmgruber commented on include/alpaka/mem/buf/BufCpu.hpp:
```diff
@@ -45,12 +45,12 @@ namespace alpaka
-            ALPAKA_FN_HOST BufCpuImpl(DevCpu const& dev, TExtent const& extent)
-                : AllocCpuAligned<std::integral_constant<std::size_t, core::vectorization::defaultAlignment>>()
-                , m_dev(dev)
+            template<typename TExtent, typename Deleter>
+            ALPAKA_FN_HOST BufCpuImpl(DevCpu const& dev, TElem* pMem, Deleter deleter, TExtent const& extent)
```
Since you are stuffing the deleter into a `std::function` later anyway, I would just use it in the constructor interface as well. But that's just personal taste.
```suggestion
            template<typename TExtent>
            ALPAKA_FN_HOST BufCpuImpl(DevCpu const& dev, TElem* pMem, std::function<void(TElem*)> deleter, TExtent const& extent)
```

fwyzard replied:
Sure, I can make the change in the `BufCpuImpl` constructor.

I would not make the change for the `BufUniformCudaHipRt` constructor, since the latter passes the deleter to a `shared_ptr`, and I don't know how that's handled internally.

And I would not make the change for the `BufCpu` constructor, to keep the interface as close as possible among the different buffers.

bernhardmgruber replied:
OK, fine for me!

bernhardmgruber commented on include/alpaka/mem/buf/BufCpu.hpp:
```diff
@@ -111,9 +102,10 @@ namespace alpaka
-        template<typename TExtent>
-        ALPAKA_FN_HOST BufCpu(DevCpu const& dev, TExtent const& extent)
-            : m_spBufCpuImpl(std::make_shared<detail::BufCpuImpl<TElem, TDim, TIdx>>(dev, extent))
+        template<typename TExtent, typename Deleter>
+        ALPAKA_FN_HOST BufCpu(DevCpu const& dev, TElem* pMem, Deleter deleter, TExtent const& extent)
+            : m_spBufCpuImpl{
+                std::make_shared<detail::BufCpuImpl<TElem, TDim, TIdx>>(dev, pMem, std::move(deleter), extent)}
```
```suggestion
        template<typename TExtent>
        ALPAKA_FN_HOST BufCpu(DevCpu const& dev, TElem* pMem, std::function<void(TElem*)> deleter, TExtent const& extent)
            : m_spBufCpuImpl{
                std::make_shared<detail::BufCpuImpl<TElem, TDim, TIdx>>(dev, pMem, std::move(deleter), extent)}
```

bernhardmgruber commented on include/alpaka/mem/buf/BufCpu.hpp:
```diff
@@ -233,7 +225,39 @@ namespace alpaka
                 ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
 
-                return BufCpu<TElem, TDim, TIdx>(dev, extent);
+                // alpaka::AllocCpuAligned is stateless
+                using Allocator
+                    = AllocCpuAligned<std::integral_constant<std::size_t, core::vectorization::defaultAlignment>>;
```
Should we assert this? Something like:
```c++
struct Test : Allocator { int i; };
static_assert(sizeof(Test) == sizeof(int));
```

fwyzard replied:
I see, but I don't think it would be useful, because a similar `assert()` would still pass for an `Allocator` that holds _e.g._ a pointer to an internal buffer, etc.

fwyzard replied:
Ah, no, sorry, I misread the example.

fwyzard replied:
What about
```c++
static_assert(std::is_empty_v<Allocator>, "AllocCpuAligned is expected to be stateless");
```
?

bernhardmgruber replied:
Perfect! I did not know about that trait :) Thx!

bernhardmgruber reviewed (commented):
Looks mostly good! Here are some more comments.

bernhardmgruber reviewed (changes requested):
Please apply the changed discussed in the various comments. I would also like to have at least 1 unit test for the proposed feature. Otherwise the PR looks good to me.

fwyzard commented:
Can you suggest one existing unit test to copy from and modify ?

bernhardmgruber commented:
> Can you suggest one existing unit test to copy from and modify ?

You can start by copying `test/unit/mem/buf/src/BufTest.cpp:22` and `...:44` and replacing `allocBuf` with `allocBufAsync`.

fwyzard commented:
I think I have addressed all the comments.
I'll be looking at adding a unit test next.

fwyzard committed:
Add unit tests for queue-ordered memory buffers

Make sure the memory buffers go out of scope and are destroyed, thus
enqueueing the call to free the memory, before waiting for the queue.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -41,6 +41,35 @@ static auto testBufferMutable(alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>>
+    // memory will be deallocated once the queue reaches this point
+    alpaka::wait(queue);
+}
+
```
@bernhardmgruber did you have in mind something like this, or something else ?

Given that queue-ordered memory allocations are available only for CUDA and CPU buffers, I expect that the tests will fail for the other accelerators.
What would be a good way to make the new test run conditionally on the accelerator type ?

Also, is there a way to run the tests locally ?
I couldn't find any instructions for it.

bernhardmgruber replied:
Something of that size looks fine to me for starters, thank you! I think you could add more tests on the lifetime aspect of these async buffers, like enqueuing a kernel, then destroying the buffer instance, then waiting for the kernel that still uses that buffer.

> What would be a good way to make the new test run conditionally on the accelerator type ?

Template specialization or overloading. I would pick the latter, add the `Acc` as unused first argument, then add empty overloads of `testAsyncBufferMutable` for the specific accelerators where the test does not work yet.

> Also, is there a way to run the tests locally ?

alpaka uses cmake's standard testing facilities. Set `BUILD_TESTING` to `ON` in cmake and build all the test targets (or just all targets since the examples should compile as well, in case you turned them on). Run `ctest` in the build dir to execute all tests.

fwyzard commented:
OK, the conditions in the test should now be complete.

bernhardmgruber commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -18,6 +18,91 @@
+    if constexpr(std::is_same_v<alpaka::Dev<TAcc>, alpaka::DevCudaRt>)
+    {
+#    if BOOST_LANG_CUDA < BOOST_VERSION_NUMBER(11, 2, 0)
+        std::cout << "Support for stream-ordered memory buffers require CUDA 11.2 or higher. "
+                  << "The test will be skipped." << std::endl;
```
```suggestion
        std::cout << "Support for stream-ordered memory buffers requires CUDA 11.2 or higher. "
                  << "The test will be skipped." << std::endl;
```

fwyzard replied:
Thanks.

bernhardmgruber commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -64,6 +178,40 @@ TEMPLATE_LIST_TEST_CASE("memBufZeroSizeTest", "[memBuf]", alpaka::test::TestAccs
+    }
+    else
+    {
+        printAsyncBufferUnsupported<Acc>();
```
```suggestion
        INFO("Stream-ordered memory buffers not supported");
```
And delete `printAsyncBufferUnsupported`. I think this is good enough. But do as you like.

Two more occurences below.

fwyzard replied:
OK for me.

bernhardmgruber commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -18,6 +18,91 @@
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+    if constexpr(std::is_same_v<alpaka::Dev<TAcc>, alpaka::DevCudaRt>)
+    {
+#    if BOOST_LANG_CUDA < BOOST_VERSION_NUMBER(11, 2, 0)
+        return false;
+#    else
+        if constexpr(alpaka::Dim<TAcc>::value > 1)
+        {
+            return false;
+        }
+#    endif // BOOST_LANG_CUDA < BOOST_VERSION_NUMBER(11, 2, 0)
```
```suggestion
        return (BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)) && alpaka::Dim<TAcc>::value == 1;
```

bernhardmgruber commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -18,6 +18,91 @@
+        std::cout << "Support for stream-ordered memory buffers require CUDA 11.2 or higher. "
+                  << "The test will be skipped." << std::endl;
+#    else
+        if constexpr(std::is_same_v<alpaka::Dev<TAcc>, alpaka::DevCudaRt> and alpaka::Dim<TAcc>::value > 1)
```
```suggestion
        if constexpr(std::is_same_v<alpaka::Dev<TAcc>, alpaka::DevCudaRt> && alpaka::Dim<TAcc>::value > 1)
```
The alternative operator keywords are only availabel in MSVC when compiling in standard conformance mode (`/permissive-`). Unfortunately, CMake chooses the non-conforming mode by default. Please avoid them for the sake of broader MSVC support.

fwyzard replied:
sigh ...

fwyzard commented:
OK, making progress with the tests, now there are only two failures left.

---
LeakSanitizer complains about
```
28: Direct leak of 31680 byte(s) in 1 object(s) allocated from:
28:     #0 0x4e12f8  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x4e12f8)
28:     #1 0x875676  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x875676)
28:     #2 0x61683d  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x61683d)
28:     #3 0x8f01e9  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8f01e9)
28:     #4 0x8d7b29  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8d7b29)
28:     #5 0x8d0e2d  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8d0e2d)
28:     #6 0x8ce7b7  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8ce7b7)
28:     #7 0x8e2a98  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8e2a98)
28:     #8 0x8e01b3  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x8e01b3)
28:     #9 0x93065d  (/home/runner/work/alpaka/alpaka/build/test/unit/mem/buf/memBufTest+0x93065d)
28:     #10 0x7fbb31ae8bf6  (/lib/x86_64-linux-gnu/libc.so.6+0x21bf6)
...
```
which I could use some help to investigate; in particular:
  - can you make the sanitizer produce a more meaningful stack trace ?
  - can you suggest how to reproduce it locally ?

---
All NVCC builds on Windows fail with
```
D:/a/alpaka/alpaka/test/unit/mem/buf/src/BufTest.cpp(72): error : expected a ")" [D:\a\alpaka\alpaka\build\test\unit\mem\buf\memBufTest.vcxproj]
             detected during:
               instantiation of "auto printAsyncBufferUnsupported<TAcc>()->void [with TAcc=alpaka::test::detail::AccGpuUniformCudaHipRtIfAvailableElseInt<alpaka::DimInt<2ULL>, uint64_t>]"
   (195): here
               instantiation of "void ____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____F_U_N_C____7<TestType>() [with TestType=alpaka::test::detail::AccGpuUniformCudaHipRtIfAvailableElseInt<alpaka::DimInt<2ULL>,
   (181): here
               instantiation of "void <unnamed>::ns_____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____6::____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____6<Types...>::reg_tests() [with Types=<alpaka::test::detail::
   (181): here
```
That code looks fine to me, and GCC and clang do not complain about it... is there some limitation with Visual Studio 2019 in C++ 17 mode ?

fwyzard commented:
OK, all checks finally pass.

The last problem was interesting, as it highlights a detail of how the stream-ordered buffers should be used: since the buffer will queue the call to free the memory when it goes out of scope, to be able to wait for it to happen the lifetime of the queue should be longer than the buffer's own scope.
This is unlikely to be an issue in a larger program, where the queue will likely be used across many functions (otherwise it makes little sense to use an async buffer), but it needed special attention in the unit test.

bernhardmgruber commented:
> since the buffer will queue the call to free the memory when it goes out of scope, to be able to wait for it to happen the lifetime of the queue should be longer than the buffer's own scope.

I wonder whether we could check that. Or even block in the queue's destructor. Or extent the queue's lifetime. Since we have some `shared_ptr`s in the backend, we could keep `shared_ptr`s or possibly `weak_ptr`s to other alpaka objects around. Like having a shared pointer to the queue an async buffer was created on inside the buffer, so the queue is kept alive, etc. Or the queue having a vector of weak pointes to buffers, to verify that they are all gone and block otherwise. But this is definitely out of the scope for this PR.

fwyzard commented:
The queue itself should already live at least as long as the buffer, since the buffer's destructor holds a copy of it - but it goes away when the call to free the memory is enqueued.

We could keep the queue alive until all pending memory operations have completed by passing a copy of it to the enqueued function itself:
```patch
diff --git a/include/alpaka/mem/buf/BufCpu.hpp b/include/alpaka/mem/buf/BufCpu.hpp
index 1d7b72e44b84..52b28400f1c6 100644
--- a/include/alpaka/mem/buf/BufCpu.hpp
+++ b/include/alpaka/mem/buf/BufCpu.hpp
@@ -261,7 +261,17 @@ namespace alpaka
                 TElem* memPtr
                     = alpaka::malloc<TElem>(Allocator{}, static_cast<std::size_t>(extent::getExtentProduct(extent)));
                 auto deleter = [queue = std::move(queue)](TElem* ptr) mutable
-                { alpaka::enqueue(queue, [ptr]() { alpaka::free(Allocator{}, ptr); }); };
+                {
+                    alpaka::enqueue(
+                        queue,
+                        [ptr, queue]()
+                        {
+                            // free the memory
+                            alpaka::free(Allocator{}, ptr);
+                            // keep the queue alive until all memory operations have cimpleted
+                            alpaka::ignore_unused(queue);
+                        });
+                };
 
                 return BufCpu<TElem, TDim, TIdx>(dev, memPtr, std::move(deleter), extent);
             }
```

I haven't considered if this may cause any circular references or not, or other issues.


Alternatively, making the queue's destructor wait on all pending operations would solve the issue, as well.
